File: bindings/python/python/maplibre_native/camera.py
"""Camera descriptors and map camera operations."""

from __future__ import annotations

from dataclasses import dataclass

from .geo import LatLng, LatLngBounds


@dataclass(frozen=True, slots=True)
class ScreenPoint:
    """Screen-space point in logical map pixels."""

    x: float
    y: float


@dataclass(frozen=True, slots=True)
class EdgeInsets:
    """Screen-space inset in logical map pixels."""

    top: float = 0.0
    left: float = 0.0
    bottom: float = 0.0
    right: float = 0.0


@dataclass(frozen=True, slots=True)
class UnitBezier:
    """Cubic Bezier control points for animation easing."""

    p1x: float
    p1y: float
    p2x: float
    p2y: float


@dataclass(frozen=True, slots=True)
class Vec3:
    """Three-component vector used by free camera options."""

    x: float
    y: float
    z: float


@dataclass(frozen=True, slots=True)
class Quaternion:
    """Quaternion stored as x, y, z, w components."""

    x: float
    y: float
    z: float
    w: float


@dataclass(frozen=True, slots=True)
class CameraOptions:
    """Camera fields used for snapshots and camera commands."""

    center: LatLng | None = None
    zoom: float | None = None
    bearing: float | None = None
    pitch: float | None = None
    padding: EdgeInsets | None = None
    anchor: ScreenPoint | None = None

    @classmethod
    def from_native(cls, raw: dict[str, object]) -> "CameraOptions":
        """Build camera options from private native bridge values."""
        center = raw["center"]
        padding = raw["padding"]
        anchor = raw["anchor"]
        return cls(
            center=LatLng(**center) if isinstance(center, dict) else None,
            zoom=raw["zoom"],
            bearing=raw["bearing"],
            pitch=raw["pitch"],
            padding=EdgeInsets(**padding) if isinstance(padding, dict) else None,
            anchor=ScreenPoint(**anchor) if isinstance(anchor, dict) else None,
        )


@dataclass(frozen=True, slots=True)
class AnimationOptions:
    """Optional animation controls for camera transitions."""

    duration_ms: float | None = None
    velocity: float | None = None
    min_zoom: float | None = None
    easing: UnitBezier | None = None


@dataclass(frozen=True, slots=True)
class CameraFitOptions:
    """Optional fitting controls for camera-for-viewport queries."""

    padding: EdgeInsets | None = None
    bearing: float | None = None
    pitch: float | None = None


@dataclass(frozen=True, slots=True)
class BoundOptions:
    """Optional map camera constraint fields."""

    bounds: LatLngBounds | None = None
    min_zoom: float | None = None
    max_zoom: float | None = None
    min_pitch: float | None = None
    max_pitch: float | None = None


@dataclass(frozen=True, slots=True)
class FreeCameraOptions:
    """Free camera position and orientation in MapLibre camera space."""

    position: Vec3 | None = None
    orientation: Quaternion | None = None


@dataclass(frozen=True, slots=True)
class ProjectionMode:
    """Axonometric rendering options for the live map render transform."""

    axonometric: bool | None = None
    x_skew: float | None = None
    y_skew: float | None = None


__all__ = [
    "AnimationOptions",
    "BoundOptions",
    "CameraFitOptions",
    "CameraOptions",
    "EdgeInsets",
    "FreeCameraOptions",
    "ProjectionMode",
    "Quaternion",
    "ScreenPoint",
    "UnitBezier",
    "Vec3",
]

File: bindings/dart/lib/src/camera/camera.dart
/// Camera descriptor and map camera operation types.
library;

import '../geo/geo.dart';

export '../geo/geo.dart'
    show EdgeInsets, LatLng, LatLngBounds, Quaternion, ScreenPoint, Vec3;

/// Cubic easing curve for animated camera transitions.
final class UnitBezier {
  /// Creates a cubic unit bezier.
  const UnitBezier(this.x1, this.y1, this.x2, this.y2);

  /// First control point x.
  final double x1;

  /// First control point y.
  final double y1;

  /// Second control point x.
  final double x2;

  /// Second control point y.
  final double y2;

  @override
  bool operator ==(Object other) =>
      other is UnitBezier &&
      other.x1 == x1 &&
      other.y1 == y1 &&
      other.x2 == x2 &&
      other.y2 == y2;

  @override
  int get hashCode => Object.hash(x1, y1, x2, y2);
}

/// Camera fields used for snapshots and camera commands.
final class CameraOptions {
  /// Creates camera options.
  const CameraOptions({
    this.center,
    this.zoom,
    this.bearing,
    this.pitch,
    this.centerAltitude,
    this.padding,
    this.anchor,
    this.roll,
    this.fieldOfView,
  });

  /// Optional center coordinate.
  final LatLng? center;

  /// Optional zoom level.
  final double? zoom;

  /// Optional bearing in degrees.
  final double? bearing;

  /// Optional pitch in degrees.
  final double? pitch;

  /// Optional center altitude.
  final double? centerAltitude;

  /// Optional viewport padding.
  final EdgeInsets? padding;

  /// Optional screen anchor.
  final ScreenPoint? anchor;

  /// Optional roll in degrees.
  final double? roll;

  /// Optional field of view.
  final double? fieldOfView;
}

/// Optional animation controls for camera transitions.
final class AnimationOptions {
  /// Creates animation options.
  const AnimationOptions({
    this.durationMs,
    this.velocity,
    this.minZoom,
    this.easing,
  });

  /// Optional duration in milliseconds.
  final double? durationMs;

  /// Optional average fly-to velocity in screenfuls per second.
  final double? velocity;

  /// Optional peak zoom for fly-to transitions.
  final double? minZoom;

  /// Optional easing curve.
  final UnitBezier? easing;
}

/// Optional fitting controls for camera-for-viewport queries.
final class CameraFitOptions {
  /// Creates camera fit options.
  const CameraFitOptions({this.padding, this.bearing, this.pitch});

  /// Optional viewport padding.
  final EdgeInsets? padding;

  /// Optional bearing in degrees.
  final double? bearing;

  /// Optional pitch in degrees.
  final double? pitch;
}

/// Optional map camera constraint fields.
final class BoundOptions {
  /// Creates bound options.
  const BoundOptions({
    this.bounds,
    this.minZoom,
    this.maxZoom,
    this.minPitch,
    this.maxPitch,
  });

  /// Optional coordinate bounds.
  final LatLngBounds? bounds;

  /// Optional minimum zoom.
  final double? minZoom;

  /// Optional maximum zoom.
  final double? maxZoom;

  /// Optional minimum pitch.
  final double? minPitch;

  /// Optional maximum pitch.
  final double? maxPitch;
}

/// Free camera position and orientation in MapLibre Native camera space.
final class FreeCameraOptions {
  /// Creates free camera options.
  const FreeCameraOptions({this.position, this.orientation});

  /// Optional position.
  final Vec3? position;

  /// Optional orientation.
  final Quaternion? orientation;
}

/// MapLibre axonometric rendering options used for snapshots and commands.
final class ProjectionModeOptions {
  /// Creates projection mode options.
  const ProjectionModeOptions({this.axonometric, this.xSkew, this.ySkew});

  /// Optional axonometric mode switch.
  final bool? axonometric;

  /// Optional native x-skew factor.
  final double? xSkew;

  /// Optional native y-skew factor.
  final double? ySkew;
}
